### ISSUE TITLE
`GetAllStates` relies on the the unreliable state string instead of the history

### ISSUE THREAD

**Issue by gabrfarina** [deepmind/open_spiel#399]
Hit as part of #376 .

+@lanctot +@michalsustr +@finbarrtimbers

The implementation of `GetAllStates` expects that `state->ToString()` be unique for states, which is not true. For example, in the Battleship game, `ToString()` returns an ascii art representation of the player boards. In Go, my understanding is that `ToString()` renders the Go board (**edit** see also [this](https://github.com/deepmind/open_spiel/pull/376#discussion_r496431791) discussion). Neither of these mappings are injective: different orders of moves can result in the same boards, and therefore the same `ToString()`.

https://github.com/deepmind/open_spiel/blob/de2c9ef57038829c2709becb0104e3b2d76725cc/open_spiel/algorithms/get_all_states.cc#L47-L53

This results in the method failing to explore the full game tree. Since the method is performing a DFS on a tree, it is impossible to hit the same state twice, so I am not sure what the comment `add only if not already present` was supposed to catch. I suggest the following:
- If we want to keep the map, we should have histories as key.
- There is no need to keep an expensive map: a vector of (unique) pointers to states would be more efficient and simplify the code even further. I'm not sure how breaking a change that would be though?

**Comment by lanctot** [comment]
That depends how you define state. :) 

The `ToString` representation is actually unique across states. If the order of moves is important for determining the state, it should also be in the `ToString`. 

I don't see how the quoted code is problematic? The traversal of the game tree still happens further below it. 

If you want a similar method bases histories, we can have another one called `GetAllHistories` but I think in your case of battleship we should just add all the shot info to the ToString (see e.g. GoofspielState::ToString). 

The ToString needs to be the complete description of the state. Having a display of the board is good: please keep it! Just add the other information necessary to describe the full state.

**Comment by finbarrtimbers** [comment]
I do think that adding a `GetAllHistories` that uses the optimization you describe would be great, and changing `count_all_states` to use it would be a good change.

**Comment by finbarrtimbers** [comment]
However, I think Marc's correct.

**Comment by gabrfarina** [comment]
> The `ToString` representation is actually unique across states. If the order of moves is important for determining the state, it should also be in the `ToString`.

OK that was also my worry when writing the generator. However, because of [this](https://github.com/deepmind/open_spiel/pull/376#discussion_r496431791) discussion I ended up throwing away `ToPrettyString` and replacing `ToString` with that implementation.

Does this mean that I should go back to having two methods? Also, @michalsustr  was suggesting that in Go `ToString` only prints the board?

> If you want a similar method based on histories, we can have another one called GetAllHistories but I think in your case of battleship we should just add all the shot info to the ToString (see e.g. GoofspielState::ToString).

I think this is problematic for the code that uses `GetAllStates` to count then number of infostates etc (for example inside of open_spiel/examples).

**Comment by lanctot** [comment]
I say just have it all in ToString. 

Why do you think it would cause a problem counting info states? The recursion doesn't stop just because there's a duplicate; it just does not add it to the hash set.

**Comment by gabrfarina** [comment]
> Why do you think it would cause a problem counting info states? The recursion doesn't stop just because there's a duplicate; it just does not add it to the hash set.

You are right, this does not actually stop iterating, my bad. But because of the fact that keys collide, the map contains less states than it should. Entire infosets get skipped. I hit this while writing the test that was checking the number of infosets & sequences of the generated Battleship games. Because of the key collisions we end up significantly underreporting the number of infosets and sequences.

**Comment by gabrfarina** [comment]
Also, I took a quick look at the implementation of Go, and it seems that indeed `ToString` only prints the board like Michal  was saying.

https://github.com/deepmind/open_spiel/blob/de2c9ef57038829c2709becb0104e3b2d76725cc/open_spiel/games/go.cc#L147-L153

**Comment by gabrfarina** [comment]
I can see three ways forward:
- Reverting the implementation of `ToString()` for Battleship to what it was before basically, and opening an issue for other games (like Go, it seems) that might not be conforming to the fact that `ToString` be a bijection between strings and states. In the meantime I can also add a noisy check that the call to `map.insert` results in a new entry as it should do. If we detect a collision we can abort with a `SpielFatalError`.
- Changing the implementation of `GetAllStates` so that it returns a map `history -> State*`
- Have `GetAllStates` return a vector of pointers instead of a map, bypassing the `ToState` vs `History` decision altogether

**Comment by lanctot** [comment]
We can't change GetAllStates at this point, too much code depends on it and it'd be a severely breaking change. 

It will work perfectly with a fixed ToString. Why is that not your preference?

(Sorry if I'm missing something)

**Comment by lanctot** [comment]
Most of the perfect information games have the current board state as the ToString, and it makes sense there because they are uniquely identified by their board representations (the history is irrelevant,  due to i.e. the Markov property). Those are used as keys for the value iteration methods, where you definitely don't want to use histories.

It's only important for the ToString to have this additional information in the imperfect information games due to the use cases. Changing to ToString in Battleship to contain both the pretty board and all other information seems like the simplest solution to this and makes it more consistent with the other games.

**Comment by gabrfarina** [comment]
> Most of the perfect information games have the current board state as the ToString, and it makes sense there because they are uniquely identified by their board representations (the history is irrelevant, due to i.e. the Markov property). Those are used as keys for the value iteration methods, where you definitely don't want to use histories.

@lanctot Ah, this is very useful. I think I finally understand. So `State` is not the same as "node in the game tree" in the OpenSpiel model. I had completely missed that, and my worry was that there were games not conforming to the standard that "ToString() fully capture the state", so that if we went with the "easy solution", we would still have wrong results when using `GetAllStates` on games like Go.

Instead, it turns out that it was intended for `ToString()` to not uniquely represent the node tree. This is a bit surprising though, it basically means that `GetAllStates` cannot be used to reliably compute the dimensions of the game tree?

**Comment by lanctot** [comment]
Ah yes perfect, I was just about to reply: 

> Instead, it turns out that it was intended for `ToString()` to not uniquely represent the node tree. This is a bit surprising though, it basically means that `GetAllStates` cannot be used to reliably compute the dimensions of the game tree?

Indeed, I think I understand now. It is this code here:

I think I understand better now, this: https://github.com/deepmind/open_spiel/blob/de2c9ef57038829c2709becb0104e3b2d76725cc/open_spiel/examples/count_all_states.cc#L42-L48

that is currently broken in perfect information games because the info states are not retrieved for all ways that they could have been reached. So let's leave this issue open, we should basically just collect the info states with a custom DFS.

**Comment by lanctot** [comment]
And, apologies: I admit that this is not necessarily well-documented, and have noted to fix this. Where do you suggest would be the best place to put this? What would have prevented this?

**Comment by gabrfarina** [comment]
> that is currently broken in perfect information games because the info states are not retrieved for all ways that they could have been reached.

Yes, exactly, that code is broken whenever `ToString` is not a one to one mapping with a node.

> And, apologies: I admit that this is not necessarily well-documented, and have noted to fix this. Where do you suggest would be the best place to put this? What would have prevented this?

No problem, glad we could get to the bottom of this :-)

I think the biggest pain point is this:
* A `State` **object** clearly represents a node in the game tree.
* However, a state **string** (that is, `state.ToString()`) does not *need to* match that, *but only in some cases*. This distinction was not clear to me, and the fact that nobody objected to the discussion about the implementation of `ToString` in Battleship, it is easy to overlook.
* This also means that just saying "state" is ambiguous: the state object is a tree node, the corresponding "state" string *might* not be. So code that uses "states" might want to be very explicit about what is meant with that. For instance, when the example code counts "terminal states", it should perhaps clarified that that means "terminal state strings". (Note that for that to make sense, it is required that all nodes with the same `ToString()` be either all terminal or all non-terminal, and that property should also somehow be enforced to avoid returning wrong results.)

Based on the previous discussion with Michal, and given that Go implements `ToString` as just the current board, I used to believe that ToString actually was just a pretty-printing convenience method that was never actually used as a key. That's also why the title of this issue says "unreliable state string" --- I just thought it was a convenience function for printing.

**Comment by gabrfarina** [comment]
If we agree that in imperfect information games there has to be a 1:1 mapping between state objects and state strings, a good bandaid here is to check that the insertion in the map here 

https://github.com/deepmind/open_spiel/blob/de2c9ef57038829c2709becb0104e3b2d76725cc/open_spiel/algorithms/get_all_states.cc#L51

be successful (that is, a new entry is created), and abort if not.

**Comment by lanctot** [comment]
Alright, I'll definitely clarify some of docs based on this discussion. 

But: the state object is not really a node in a game tree. The node in the game tree is determined by a state's history. The state itself is the collection of information describing the state, which should be exposed in string form via ToString. What defines the state is naturally game type-specific. 

For example, in pig or oshi-zumo, the game state corresponds to the state in the MDP sense, which makes sense, because they are both cyclical and Markovian (perfect information). But it's not the game's string representation that dictates this, it's the type of game. The string is just a representation of the state that is useful to use as keys.

Apologies for the conflicting advice and missing that discussion.

**Comment by lanctot** [comment]
@finbarrtimbers assignment this to you. See discussion above. 

I think we should modify `count_all_states.cc`, one of these options (you choose):

- restrict `count_all_states.cc` to imperfect information games (have a check at the start of it)
- or remove the counting of information states if it's not an imperfect information game
- write a custom tree traversal rather than use `GetAllStates`

**Comment by michalsustr** [comment]
> Based on the previous discussion with Michal [..] I used to believe that ToString [..] was never actually used as a key.

I used to believe that as well O:-)

> the state object is a tree node

This a slightly confusing misconception, I agree with what Marc wrote. It just so happens that `State` contains history information as it is often practically useful. Some time ago I wrote down a [few notes about this](https://github.com/deepmind/open_spiel/tree/master/open_spiel/fog). I think the current behaviour of `GetAllStates` is correct: It collects all the underlying _world states_ (in the FOG formalism), which != game tree node (history). I think if we want to count all histories, we should do exactly that - `GetAllHistories().size()` or something like that, like Marc suggested (DFS traversal).

> So, it is indeed this code here [..] open_spiel/open_spiel/examples/count_all_states.cc [..] is currently broken 

Yes, indeed it is broken, it should use `GetAllHistories()`.

> I think we should modify examples/count_all_states.cc : [..] write a custom tree traversal rather than use GetAllStates

+1 for this, i.e. adding a `std::vector<std::unique_ptr<State>> GetAllHistories()`. 

---

Additionally, I would suggest that we rename `GetAllStates` to `GetAllWorldStates` as the name is more suggestive of what it does, and make the former deprecated.
